modules/runner/src/main/java/org/apache/ignite/internal/storage/LocalConfigurationStorage.java

```diff
@@ -49,47 +70,122 @@ public LocalConfigurationStorage(VaultManager vaultMgr) {
+
+        int i = 0;
+
+        latch = new CountDownLatch(newValues.size());
+
         for (Map.Entry<String, Serializable> entry : newValues.entrySet()) {
+            ByteArray key = ByteArray.fromString(LOC_PREFIX + "." + entry.getKey());
+
             if (entry.getValue() != null)
-                map.put(entry.getKey(), entry.getValue());
+                futs[i++] = vaultMgr.put(key, ByteUtils.toBytes(entry.getValue()));
             else
-                map.remove(entry.getKey());
+                futs[i++] = vaultMgr.remove(key);
         }
 
-        this.version.incrementAndGet();
+        try {
+            CompletableFuture.allOf(futs).get();
+
+            latch.await();
+
+            for (Map.Entry<String, Serializable> entry : newValues.entrySet()) {
+                ByteArray key = ByteArray.fromString(LOC_PREFIX + "." + entry.getKey());
+
+                Entry e = vaultMgr.get(key).get();
 
-        listeners.forEach(listener -> listener.onEntriesChanged(new Data(newValues, this.version.get(), 0)));
+                if (Arrays.equals(e.value(), ByteUtils.toBytes(entry.getValue())))
+                    // value by some key was overwritten, that means that changes not
+                    // from LocalConfigurationStorage.write overlapped with current changes, so write should be retried.
+                    return CompletableFuture.completedFuture(false);
+            }
+        }
+        catch (InterruptedException | ExecutionException e) {
+            return CompletableFuture.failedFuture(e);
+        }
 
         return CompletableFuture.completedFuture(true);
     }
 
     /** {@inheritDoc} */
-    @Override public void addListener(ConfigurationStorageListener listener) {
-        listeners.add(listener);
+    @Override public void addListener(ConfigurationStorageListener lsnr) {
+        listeners.add(lsnr);
+
+        if (watchId == 0) {
+            try {
+                watchId = vaultMgr.watch(new VaultWatch(locKeysStartRange, locKeysEndRange, new VaultListener() {
+                    // In the current implementation entries always contains only one entry
+                    @Override public boolean onUpdate(@NotNull Iterable<Entry> entries) {
+                        HashMap<String, Serializable> data = new HashMap<>();
+
+                        for (Entry e : entries) {
+                            data.put(e.key().toString().substring(locKeysEndRange.toString().length()),
+                                (Serializable)ByteUtils.fromBytes(e.value()));
+                        }
+
+                        listeners.forEach(listener -> listener.onEntriesChanged(new Data(data, ver.incrementAndGet())));
+
+                        latch.countDown();
+
+                        return true;
+                    }
+
+                    @Override public void onError(@NotNull Throwable e) {
+                        LOG.error("Vault listener issue", e);
```

Comment:
Same as for distributedConfigurationStorage.

Reply:
this code was removed at all